src/NUnitFramework/framework/Internal/Execution/SimpleWorkItemDispatcher.cs

```diff
@@ -59,17 +59,17 @@ public void Dispatch(WorkItem work)
             if (work != null)
                 work.Execute();
 #else
-            if (_topLevelWorkItem != null)
+            if (_topLevelWorkItem != null || work.RunsOnMainThread)
```

Comment:
The `if` statement is intended to distinguish the first time any `WorkItem` is run, so that we  set its value in `_topLevelWorkItem`.  I think you need a different approach here. I'll try to outline in a top-level comment so it won't be lost as changes are made.

Reply:
I hadn't commented on implementation, as I think we need to discuss the concept itself first - but I'd prefer to see a separate IWorkItemDispatcher for this, chosen by a frameworkpackagesetting higher up. Both the netstandard builds also execute directly on the main thread - which is currently in SimpleWorkItemDispatcher with #ifdef's - separating this out into a separate implementation would be cleaner imo.

Reply:
(Charlie - apologies, your other comment hadn't displayed when I typed this! Hope we're in agreement.)

Reply:
@ChrisMaddock I think we are in fundamental agreement. Whether we have a separate dispatcher or use a setting that effects how the dispatcher works is a detail that could be decided based on how many #if statements apply.

Reply:
I looked at the code. Seems like it could be done either way. However, I think we should delay till my parallel work, which may change the `IWorkItemDispatcher` interface is done. Specifically, I'm trying to get rid of the Start method entirely, which is where one of the biggest differences is found.

Reply:
Thank you for your comments. Do you think I should wait until your changes @CharliePoole are done in 3.8 and then try to reapply my changes (with your suggestions) on top of that, in the `Start` method of `SimpleWorkItemDispatcher`?